packages/emnapi/src/value/create.ts

```diff
@@ -284,15 +284,38 @@ function napi_create_buffer (
 // @ts-expect-error
 ): napi_status {
   // eslint-disable-next-line @typescript-eslint/no-unused-vars
-  let value: number
+  let value: number, pointer: number
 
   $PREAMBLE!(env, (envObject) => {
     $CHECK_ARG!(envObject, result)
-    $from64('result')
-    const arrayBuffer = createArrayBuffer(size, data)
-    const buffer = Buffer.from(arrayBuffer)
+
+    pointer = $makeMalloc('napi_create_buffer', 'size')
+    if (!pointer) throw new Error('Out of memory')
+    $from64('size')
+    for (let i = pointer; i < pointer + size; ++i) {
+      HEAPU8[i] = 0
+    }
+    const pointerInfo: PointerInfo = {
```

Comment:
This looks a bit complicated. I think what I'd do instead is check if user requested `data` pointer.

If they didn't, then we can create regular JS `Buffer` object and not bother with `malloc`, finalizers and so on.

If they did request `data`, then 1) malloc + zero memory like you already did here and 2) delegate the rest to `_napi_create_external_buffer(...)`.

That way you don't need custom handling or separate `WeakMap`.

Reply:
`emnapi_get_memory_address` needs the separate buffer weakmap

Reply:
Why can't it use the same info it uses for other TypedArrays? As in, `Buffer` is just a typed array, so it can store the same descriptor in the same `WeakMap` as other typed arrays, and the return address will be just `byteOffset`.

Reply:
In addition to telling the user the address of the pointer, `emnapi_get_memory_address` also needs to tell the user who manages the memory and whether it is allocated by emnapi, so that users can manually free memory in an environment without FinalizationRegistry.

> Why can't it use the same info it uses for other TypedArrays?

This question make me aware of another problem.

`napi_create_external_buffer` and `emnapi_create_memory_view` do not know whether the `data` is allocated by user or emnapi, so the buffer created by `napi_create_buffer` need to the new WeakMap to save this info, currently views created by `emnapi_create_memory_view` [always treated as managed by the user](https://github.com/toyobayashi/emnapi/blob/9ea15a36b29e13a7200452588078e29ac12888e2/packages/emnapi/src/value/convert2c.ts#L152).

Reply:
I removed the buffer WeakMap, but I have no idea how to solve the problem mentioned before.

> views created by emnapi_create_memory_view always treated as managed by the user.

user may create an ArrayBuffer by using `napi_create_arraybuffer`, then using `emnapi_create_memory_view` or `napi_create_external_buffer` to create a view from this arraybuffer, pass the view to `emnapi_get_memory_address` will return wrong `ownership` and `runtime_allocated`.

Reply:
Hm, isn't it enough to check if the ArrayBuffer / ArrayBufferView is in the WeakMap? If it is, then you know it's one created by emnapi, if it's not, then it's a user object and you can skip any memory sync & memory mapping.

Reply:
That is, I think it's reasonable to document that `emnapi_get_memory_address` and similar functions just won't work on JS-land ArrayBuffers, but all the `napi_*` functions should.

Reply:
Not all ArrayBuffer / ArrayBufferView in WeakMap is allocated by emnapi, for example, `napi_create_external_arraybuffer` is also don't know where the `data` is from, but the created arraybuffer should be in the WeakMap because it needs sync.

I will find my spare time to think about how to write document related with arraybuffer and views.